src/modules/mc_pos_control/MulticopterPositionControl.cpp

```diff
@@ -289,27 +277,115 @@ void MulticopterPositionControl::Run()
+			_trajectory_setpoint_sub.update(&_setpoint);
 
-			// check if all local states are valid and map accordingly
-			set_vehicle_states(setpoint.vz);
-			_control.setState(_states);
+			// adjust existing (or older) setpoint with any EKF reset deltas
+			if (_setpoint.timestamp < local_pos.timestamp) {
+				if (local_pos.vxy_reset_counter != _vxy_reset_counter) {
+					_setpoint.vx += local_pos.delta_vxy[0];
+					_setpoint.vy += local_pos.delta_vxy[1];
+				}
 
-			vehicle_constraints_s constraints;
+				if (local_pos.vz_reset_counter != _vz_reset_counter) {
+					_setpoint.vz += local_pos.delta_vz;
+				}
 
-			if (_vehicle_constraints_sub.update(&constraints)) {
-				_control.setConstraints(constraints);
-				_control.setThrustLimits(constraints.minimum_thrust, _param_mpc_thr_max.get());
+				if (local_pos.xy_reset_counter != _xy_reset_counter) {
+					_setpoint.x += local_pos.delta_xy[0];
+					_setpoint.y += local_pos.delta_xy[1];
+				}
 
-				if (constraints.reset_integral) {
-					_control.resetIntegral();
+				if (local_pos.z_reset_counter != _z_reset_counter) {
+					_setpoint.z += local_pos.delta_z;
 				}
+
+				if (local_pos.heading_reset_counter != _heading_reset_counter) {
+					_setpoint.yaw += local_pos.delta_heading;
+				}
+			}
+
+			// update vehicle constraints and handle smooth takeoff
+			_vehicle_constraints_sub.update(&_vehicle_constraints);
+
+			// fix to prevent the takeoff ramp to ramp to a too high value or get stuck because of NAN
+			// TODO: this should get obsolete once the takeoff limiting moves into the flight tasks
+			if (!PX4_ISFINITE(_vehicle_constraints.speed_up) || (_vehicle_constraints.speed_up > _param_mpc_z_vel_max_up.get())) {
+				_vehicle_constraints.speed_up = _param_mpc_z_vel_max_up.get();
+			}
+
+			// handle smooth takeoff
+			_takeoff.updateTakeoffState(_control_mode.flag_armed, _vehicle_land_detected.landed, _vehicle_constraints.want_takeoff,
+						    _vehicle_constraints.speed_up, false, time_stamp_now);
+
+			const bool not_taken_off = (_takeoff.getTakeoffState() < TakeoffState::rampup);
+			const bool flying = (_takeoff.getTakeoffState() >= TakeoffState::flight);
+			const bool flying_but_ground_contact = (flying && _vehicle_land_detected.ground_contact);
+
+			if (not_taken_off || flying_but_ground_contact) {
+				// we are not flying yet and need to avoid any corrections
+				reset_setpoint_to_nan(_setpoint);
+				Vector3f(0.f, 0.f, 100.f).copyTo(_setpoint.acceleration); // High downwards acceleration to make sure there's no thrust
+
+				// prevent any integrator windup
+				_control.resetIntegral();
 			}
 
+			// limit tilt during takeoff ramupup
+			if (_takeoff.getTakeoffState() < TakeoffState::flight) {
+				_control.setTiltLimit(math::radians(_param_mpc_tiltmax_lnd.get()));
+
+			} else {
+				_control.setTiltLimit(math::radians(_param_mpc_tiltmax_air.get()));
+			}
+
+			const float speed_up = _takeoff.updateRamp(dt, _vehicle_constraints.speed_up);
```

Comment:
Another thing that I seemed to have overlooked is that we have to guard against passing NAN here. See https://github.com/PX4/PX4-Autopilot/pull/14339

Reply:
Fixed in 7a0fd7e
I also tested it works ✔️